src/browsers/tob/shared/components/DotplotHeatmap.js

```diff
@@ -0,0 +1,218 @@
+  },
+}) => {
+  const xScaleLocal = useMemo(
+    () =>
+      xScale ||
+      scaleBand()
+        .range([0, innerWidth])
+        .domain([...new Set(data.map((item) => accessors.x(item)))])
+        .padding(tileSpacing),
+    [xScale, data]
+  )
+
+  const yScaleLocal = useMemo(
+    () =>
+      yScale ||
+      scaleBand()
+        .range([innerHeight, 0])
+        .domain([...new Set(data.map((item) => accessors.y(item)))])
+        .padding(tileSpacing),
+    [yScale, data]
+  )
+  return (
+    <>
+      <svg id={id} width={width} height={height}>
+        <g transform={`translate(${margin.left}, 40)`}>
+          <text> {title} </text>
+        </g>
+        <defs>
+          <pattern
+            id="missing-pattern"
+            width={4} // hatchsize as a prop?
+            height={4} // hatchsize
+            patternTransform="rotate(45)"
+            patternUnits="userSpaceOnUse"
+          >
+            <rect width={2} height={4} fill="black" opacity={0.4} />
+          </pattern>
+          <linearGradient id="linear-gradient" x1="0%" y1="0%" x2="0%" y2="100%">
+            <stop offset="0%" stopColor={`${colorScale.range()[0]}`} />
+            <stop offset="100%" stopColor={`${colorScale.range()[1]}`} />
+          </linearGradient>
+        </defs>
+        <g transform={`translate(${margin.left}, ${margin.top})`}>
+          <g id="x-axis">
+            <line x2={`${innerWidth}`} stroke="black" />
+            <line y1={`${innerHeight}`} y2={`${innerHeight}`} x2={`${innerWidth}`} stroke="black" />
+            {xScaleLocal.domain().map((tick) => (
+              <g
+                key={tick}
+                transform={`translate(${
+                  xScaleLocal(tick) + xScaleLocal.bandwidth() / 2
+                }, ${innerHeight})`}
+              >
+                <text
+                  transform="translate(0, 10)rotate(-90)"
+                  style={{ textAnchor: 'end', alignmentBaseline: 'middle' }}
+                >
+                  {tick}
+                </text>
+                <line y2={6} stroke="black" />
+                <line y2={`-${innerHeight}`} stroke="lightgrey" />
+              </g>
+            ))}
+          </g>
+          <g id="y-axis">
+            <line y2={`${innerHeight}`} stroke="black" />
+            <line x1={`${innerWidth}`} y2={`${innerHeight}`} x2={`${innerWidth}`} stroke="black" />
+            {yScaleLocal.domain().map((tick) => (
+              <g
+                key={tick}
+                transform={`translate(0, ${yScaleLocal(tick) + yScaleLocal.bandwidth() / 2})`}
+              >
+                <text key={tick} style={{ textAnchor: 'end', alignmentBaseline: 'middle' }} x={-6}>
+                  {tick}
+                </text>
+                <line x2={-3} stroke="black" />
+                <line x2={`${innerWidth}`} stroke="lightgrey" />
+              </g>
+            ))}
+          </g>
+          <g id="tiles">
+            {data.map((item) => (
```

Comment:
I never specified this initially, but in the case of an empty data array, can you render a centered text element with a message like 'No data to display'. Rendering an empty space makes it seem like something broke.